File: swift/Sources/benchmarks/MazeGenerator.swift
import Foundation

final class MazeGenerator: BenchmarkProtocol {
  enum MazeCell {
    case wall
    case path
  }

  class Maze {
    private let width: Int
    private let height: Int
    private var cells: [[MazeCell]]

    init(width: Int, height: Int) {

      let safeWidth = max(width, 5)
      let safeHeight = max(height, 5)
      self.width = safeWidth
      self.height = safeHeight
      self.cells = Array(
        repeating: Array(repeating: .wall, count: safeWidth),
        count: safeHeight)
    }

    subscript(x: Int, y: Int) -> MazeCell {
      get {
        guard x >= 0 && x < width && y >= 0 && y < height else {
          return .wall
        }
        return cells[y][x]
      }
      set {
        guard x >= 0 && x < width && y >= 0 && y < height else { return }
        cells[y][x] = newValue
      }
    }

    private func addRandomPaths() {
      let numExtraPaths = (width * height) / 20

      for _ in 0..<numExtraPaths {

        guard width > 2 && height > 2 else { continue }

        let x = Helper.nextInt(max: width - 2) + 1
        let y = Helper.nextInt(max: height - 2) + 1

        guard x > 0 && x < width - 1 && y > 0 && y < height - 1 else { continue }

        if self[x, y] == .wall && self[x - 1, y] == .wall && self[x + 1, y] == .wall
          && self[x, y - 1] == .wall && self[x, y + 1] == .wall
        {
          self[x, y] = .path
        }
      }
    }

    private func divide(x1: Int, y1: Int, x2: Int, y2: Int) {
      let width = x2 - x1
      let height = y2 - y1

      if width < 2 || height < 2 {
        return
      }

      let widthForWall = max(width - 2, 0)
      let heightForWall = max(height - 2, 0)
      let widthForHole = max(width - 1, 0)
      let heightForHole = max(height - 1, 0)

      if widthForWall == 0 || heightForWall == 0 || widthForHole == 0 || heightForHole == 0 {
        return
      }

      if width > height {

        let wallRange = max(widthForWall / 2, 1)
        let wallOffset = wallRange > 0 ? (Helper.nextInt(max: wallRange)) * 2 : 0
        let wallX = x1 + 2 + wallOffset

        let holeRange = max(heightForHole / 2, 1)
        let holeOffset = holeRange > 0 ? (Helper.nextInt(max: holeRange)) * 2 : 0
        let holeY = y1 + 1 + holeOffset

        guard wallX >= x1 + 2 && wallX <= x2 else { return }
        guard holeY >= y1 + 1 && holeY <= y2 else { return }

        for y in y1...y2 {
          if y != holeY {
            self[wallX, y] = .wall
          }
        }

        if wallX > x1 + 1 {
          divide(x1: x1, y1: y1, x2: wallX - 1, y2: y2)
        }
        if wallX + 1 < x2 {
          divide(x1: wallX + 1, y1: y1, x2: x2, y2: y2)
        }
      } else {

        let wallRange = max(heightForWall / 2, 1)
        let wallOffset = wallRange > 0 ? (Helper.nextInt(max: wallRange)) * 2 : 0
        let wallY = y1 + 2 + wallOffset

        let holeRange = max(widthForHole / 2, 1)
        let holeOffset = holeRange > 0 ? (Helper.nextInt(max: holeRange)) * 2 : 0
        let holeX = x1 + 1 + holeOffset

        guard wallY >= y1 + 2 && wallY <= y2 else { return }
        guard holeX >= x1 + 1 && holeX <= x2 else { return }

        for x in x1...x2 {
          if x != holeX {
            self[x, wallY] = .wall
          }
        }

        if wallY > y1 + 1 {
          divide(x1: x1, y1: y1, x2: x2, y2: wallY - 1)
        }
        if wallY + 1 < y2 {
          divide(x1: x1, y1: wallY + 1, x2: x2, y2: y2)
        }
      }
    }

    private func isConnected(startX: Int, startY: Int, goalX: Int, goalY: Int) -> Bool {

      guard startX >= 0 && startX < width && startY >= 0 && startY < height else {
        return false
      }
      guard goalX >= 0 && goalX < width && goalY >= 0 && goalY < height else {
        return false
      }

      var visited = Array(
        repeating: Array(repeating: false, count: width),
        count: height)
      var queue: [(x: Int, y: Int)] = [(startX, startY)]
      visited[startY][startX] = true

      while !queue.isEmpty {
        let (x, y) = queue.removeFirst()

        if x == goalX && y == goalY {
          return true
        }

        let directions = [(0, -1), (1, 0), (0, 1), (-1, 0)]

        for (dx, dy) in directions {
          let nx = x + dx
          let ny = y + dy

          guard nx >= 0 && nx < width && ny >= 0 && ny < height else {
            continue
          }

          if self[nx, ny] == .path && !visited[ny][nx] {
            visited[ny][nx] = true
            queue.append((nx, ny))
          }
        }
      }

      return false
    }

    func generate() {
      if width < 5 || height < 5 {

        let midY = height / 2
        for x in 0..<width {
          self[x, midY] = .path
        }
        return
      }

      divide(x1: 0, y1: 0, x2: width - 1, y2: height - 1)
      addRandomPaths()
    }

    func toBoolGrid() -> [[Bool]] {
      var result = Array(
        repeating: Array(repeating: false, count: width),
        count: height)
      for y in 0..<height {
        for x in 0..<width {
          result[y][x] = (self[x, y] == .path)
        }
      }
      return result
    }

    static func generateWalkableMaze(width: Int, height: Int) -> [[Bool]] {
      let maze = Maze(width: width, height: height)
      maze.generate()

      let startX = 1
      let startY = 1
      let goalX = width - 2
      let goalY = height - 2

      if width >= 3 && height >= 3 {
        if !maze.isConnected(
          startX: startX, startY: startY,
          goalX: goalX, goalY: goalY)
        {

          for x in 0..<width {
            for y in 0..<height {
              if x == 1 || y == 1 || x == width - 2 || y == height - 2 {
                maze[x, y] = .path
              }
            }
          }
        }
      }

      return maze.toBoolGrid()
    }
  }

  private var resultVal: UInt32 = 0
  private var width: Int = 0
  private var height: Int = 0
  private var boolGrid: [[Bool]] = []

  init() {

    let configW = Int(configValue("w") ?? 50)
    let configH = Int(configValue("h") ?? 50)
    self.width = max(configW, 5)
    self.height = max(configH, 5)
  }

  private func gridChecksum(_ grid: [[Bool]]) -> UInt32 {
    var hasher: UInt32 = 2_166_136_261
    let prime: UInt32 = 16_777_619

    for y in 0..<grid.count {
      let row = grid[y]
      for x in 0..<row.count {
        if row[x] {

          let x32 = UInt32(x)
          let jSquared = x32 &* x32

          hasher = hasher ^ jSquared
          hasher = hasher &* prime
        }
      }
    }
    return hasher
  }

  func run(iterationId: Int) {
    boolGrid = Maze.generateWalkableMaze(width: width, height: height)
  }

  var checksum: UInt32 {
    return gridChecksum(boolGrid)
  }

  func prepare() {

  }

  func name() -> String {
    return "MazeGenerator"
  }
}

File: zig/src/maze_generator.zig
const std = @import("std");
const Benchmark = @import("benchmark.zig").Benchmark;
const Helper = @import("helper.zig").Helper;

pub const MazeGenerator = struct {
    allocator: std.mem.Allocator,
    helper: *Helper,
    width: i32,
    height: i32,
    bool_grid: ?[]const []const bool,
    result_val: u32,

    const Cell = enum(u8) { wall = 0, path = 1 };

    const vtable = Benchmark.VTable{
        .run = runImpl,
        .checksum = resultImpl,
        .deinit = deinitImpl,
    };

    const Maze = struct {
        width: i32,
        height: i32,
        cells: []Cell,
        allocator: std.mem.Allocator,
        helper: *Helper,

        fn init(allocator: std.mem.Allocator, helper: *Helper, width: i32, height: i32) !Maze {
            const actual_width = if (width < 5) 5 else width;
            const actual_height = if (height < 5) 5 else height;
            const cells = try allocator.alloc(Cell, @as(usize, @intCast(actual_width * actual_height)));
            @memset(cells, Cell.wall);

            return Maze{
                .width = actual_width,
                .height = actual_height,
                .cells = cells,
                .allocator = allocator,
                .helper = helper,
            };
        }

        fn deinit(self: *Maze) void {
            self.allocator.free(self.cells);
        }

        fn get(self: *const Maze, x: i32, y: i32) Cell {
            if (x < 0 or x >= self.width or y < 0 or y >= self.height) {
                return Cell.wall;
            }
            return self.cells[@as(usize, @intCast(y * self.width + x))];
        }

        fn set(self: *Maze, x: i32, y: i32, cell: Cell) void {
            if (x >= 0 and x < self.width and y >= 0 and y < self.height) {
                self.cells[@as(usize, @intCast(y * self.width + x))] = cell;
            }
        }

        fn addRandomPaths(self: *Maze) void {
            const num_extra_paths = @as(i32, @intCast(@divTrunc(self.width * self.height, 20)));
            var i: i32 = 0;
            while (i < num_extra_paths) : (i += 1) {
                const x = self.helper.nextInt(self.width - 2) + 1;
                const y = self.helper.nextInt(self.height - 2) + 1;

                if (x >= 1 and x < self.width - 1 and y >= 1 and y < self.height - 1) {
                    if (self.get(x, y) == Cell.wall and
                        self.get(x - 1, y) == Cell.wall and
                        self.get(x + 1, y) == Cell.wall and
                        self.get(x, y - 1) == Cell.wall and
                        self.get(x, y + 1) == Cell.wall)
                    {
                        self.set(x, y, Cell.path);
                    }
                }
            }
        }

        fn divide(self: *Maze, x1: i32, y1: i32, x2: i32, y2: i32) void {
            const width = x2 - x1;
            const height = y2 - y1;

            if (width < 2 or height < 2) return;

            const width_for_wall = @max(width - 2, 0);
            const height_for_wall = @max(height - 2, 0);
            const width_for_hole = @max(width - 1, 0);
            const height_for_hole = @max(height - 1, 0);

            if (width_for_wall == 0 or height_for_wall == 0 or
                width_for_hole == 0 or height_for_hole == 0) return;

            if (width > height) {
                const wall_range = @max(@divTrunc(width_for_wall, 2), 1);
                const wall_offset = if (wall_range > 0) self.helper.nextInt(@as(i32, @intCast(wall_range))) * 2 else 0;
                const wall_x = x1 + 2 + wall_offset;

                const hole_range = @max(@divTrunc(height_for_hole, 2), 1);
                const hole_offset = if (hole_range > 0) self.helper.nextInt(@as(i32, @intCast(hole_range))) * 2 else 0;
                const hole_y = y1 + 1 + hole_offset;

                if (wall_x > x2 or hole_y > y2) return;

                var y = y1;
                while (y <= y2) : (y += 1) {
                    if (y != hole_y) {
                        self.set(wall_x, y, Cell.wall);
                    }
                }

                if (wall_x > x1 + 1) self.divide(x1, y1, wall_x - 1, y2);
                if (wall_x + 1 < x2) self.divide(wall_x + 1, y1, x2, y2);
            } else {
                const wall_range = @max(@divTrunc(height_for_wall, 2), 1);
                const wall_offset = if (wall_range > 0) self.helper.nextInt(@as(i32, @intCast(wall_range))) * 2 else 0;
                const wall_y = y1 + 2 + wall_offset;

                const hole_range = @max(@divTrunc(width_for_hole, 2), 1);
                const hole_offset = if (hole_range > 0) self.helper.nextInt(@as(i32, @intCast(hole_range))) * 2 else 0;
                const hole_x = x1 + 1 + hole_offset;

                if (wall_y > y2 or hole_x > x2) return;

                var x = x1;
                while (x <= x2) : (x += 1) {
                    if (x != hole_x) {
                        self.set(x, wall_y, Cell.wall);
                    }
                }

                if (wall_y > y1 + 1) self.divide(x1, y1, x2, wall_y - 1);
                if (wall_y + 1 < y2) self.divide(x1, wall_y + 1, x2, y2);
            }
        }

        fn isConnected(self: *const Maze, start_x: i32, start_y: i32, goal_x: i32, goal_y: i32) bool {
            if (start_x < 0 or start_x >= self.width or start_y < 0 or start_y >= self.height or
                goal_x < 0 or goal_x >= self.width or goal_y < 0 or goal_y >= self.height)
            {
                return false;
            }

            var visited = self.allocator.alloc(bool, @as(usize, @intCast(self.width * self.height))) catch return false;
            defer self.allocator.free(visited);
            @memset(visited, false);

            var queue = std.ArrayList([2]i32).initCapacity(self.allocator, @as(usize, @intCast(self.width * self.height))) catch return false;
            defer queue.deinit(self.allocator);

            visited[@as(usize, @intCast(start_y * self.width + start_x))] = true;
            queue.append(self.allocator, .{ start_x, start_y }) catch return false;

            var queue_index: usize = 0;

            while (queue_index < queue.items.len) {
                const current = queue.items[queue_index];
                queue_index += 1;
                const x = current[0];
                const y = current[1];

                if (x == goal_x and y == goal_y) {
                    return true;
                }

                if (y > 0 and self.get(x, y - 1) == Cell.path) {
                    const idx = @as(usize, @intCast((y - 1) * self.width + x));
                    if (!visited[idx]) {
                        visited[idx] = true;
                        queue.append(self.allocator, .{ x, y - 1 }) catch return false;
                    }
                }

                if (x + 1 < self.width and self.get(x + 1, y) == Cell.path) {
                    const idx = @as(usize, @intCast(y * self.width + (x + 1)));
                    if (!visited[idx]) {
                        visited[idx] = true;
                        queue.append(self.allocator, .{ x + 1, y }) catch return false;
                    }
                }

                if (y + 1 < self.height and self.get(x, y + 1) == Cell.path) {
                    const idx = @as(usize, @intCast((y + 1) * self.width + x));
                    if (!visited[idx]) {
                        visited[idx] = true;
                        queue.append(self.allocator, .{ x, y + 1 }) catch return false;
                    }
                }

                if (x > 0 and self.get(x - 1, y) == Cell.path) {
                    const idx = @as(usize, @intCast(y * self.width + (x - 1)));
                    if (!visited[idx]) {
                        visited[idx] = true;
                        queue.append(self.allocator, .{ x - 1, y }) catch return false;
                    }
                }
            }

            return false;
        }

        fn generate(self: *Maze) void {
            if (self.width < 5 or self.height < 5) {
                const mid_y = @divTrunc(self.height, 2);
                var x: i32 = 0;
                while (x < self.width) : (x += 1) {
                    self.set(x, mid_y, Cell.path);
                }
                return;
            }

            self.divide(0, 0, self.width - 1, self.height - 1);

            self.addRandomPaths();
        }

        fn toBoolGrid(self: *Maze) ![]const []const bool {
            const rows = try self.allocator.alloc([]bool, @as(usize, @intCast(self.height)));
            errdefer self.allocator.free(rows);

            for (0..@as(usize, @intCast(self.height))) |i| {
                const row = try self.allocator.alloc(bool, @as(usize, @intCast(self.width)));
                errdefer for (0..i) |j| self.allocator.free(rows[j]);

                for (0..@as(usize, @intCast(self.width))) |j| {
                    row[j] = self.get(@as(i32, @intCast(j)), @as(i32, @intCast(i))) == Cell.path;
                }
                rows[i] = row;
            }

            return rows;
        }
    };

    pub fn generateWalkableMaze(allocator: std.mem.Allocator, helper: *Helper, width: i32, height: i32) ![]const []const bool {
        var maze = try Maze.init(allocator, helper, width, height);
        defer maze.deinit();
        maze.generate();

        const start_x: i32 = 1;
        const start_y: i32 = 1;
        const goal_x = width - 2;
        const goal_y = height - 2;

        if (!maze.isConnected(start_x, start_y, goal_x, goal_y)) {
            var x: i32 = 0;
            while (x < width) : (x += 1) {
                var y: i32 = 0;
                while (y < height) : (y += 1) {
                    if (x == 1 or y == 1 or x == width - 2 or y == height - 2) {
                        maze.set(x, y, Cell.path);
                    }
                }
            }
        }

        return maze.toBoolGrid();
    }

    pub fn init(allocator: std.mem.Allocator, helper: *Helper) !*MazeGenerator {
        const w = helper.config_i64("MazeGenerator", "w");
        const h = helper.config_i64("MazeGenerator", "h");
        const self = try allocator.create(MazeGenerator);
        self.* = MazeGenerator{
            .allocator = allocator,
            .helper = helper,
            .width = @as(i32, @intCast(w)),
            .height = @as(i32, @intCast(h)),
            .bool_grid = null,
            .result_val = 0,
        };
        return self;
    }

    pub fn deinit(self: *MazeGenerator) void {
        if (self.bool_grid) |grid| {
            for (grid) |row| {
                self.allocator.free(row);
            }
            self.allocator.free(grid);
        }
        self.allocator.destroy(self);
    }

    pub fn asBenchmark(self: *MazeGenerator) Benchmark {
        return Benchmark.init(self, &vtable, self.helper, "MazeGenerator");
    }

    fn gridChecksum(grid: []const []const bool) u32 {
        var hasher: u32 = 2166136261;
        const prime: u32 = 16777619;
        for (grid) |row| {
            for (row, 0..) |cell, j| {
                if (cell) {
                    const j_squared = @as(u32, @intCast(j * j));
                    hasher = (hasher ^ j_squared) *% prime;
                }
            }
        }
        return hasher;
    }

    fn runImpl(ptr: *anyopaque, _: i64) void {
        const self: *MazeGenerator = @ptrCast(@alignCast(ptr));

        if (self.bool_grid) |old_grid| {
            for (old_grid) |row| {
                self.allocator.free(row);
            }
            self.allocator.free(old_grid);
        }

        if (generateWalkableMaze(self.allocator, self.helper, self.width, self.height)) |grid| {
            self.bool_grid = grid;
        } else |_| {
            self.bool_grid = null;
        }
    }

    fn resultImpl(ptr: *anyopaque) u32 {
        const self: *MazeGenerator = @ptrCast(@alignCast(ptr));
        if (self.bool_grid) |grid| {
            return gridChecksum(grid);
        }
        return 0;
    }

    fn deinitImpl(ptr: *anyopaque) void {
        const self: *MazeGenerator = @ptrCast(@alignCast(ptr));
        self.deinit();
    }
};
